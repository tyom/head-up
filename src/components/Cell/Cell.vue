<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{ _toggled: showSettings }"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h" />
      </button>
    </header>
    <div class="body-wrapper">
      <div :class="bodyClass" class="body" :style="bodyStyle">
        <slot v-if="$slots.default" />
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings" />
      </transition>
    </div>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import store from '../../store';
import { createGridTemplate } from '../../utils';
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    grid: String,
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  created() {
    if (!this.$slots.default) {
      return;
    }
    const { areas } = createGridTemplate(this.grid);
    this.$slots.default.forEach((vnode, idx) => {
      vnode.data = merge(vnode.data, {
        style: {
          gridArea: areas[idx],
        },
      });
    });
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && store.state.editMode;
    },
    bodyStyle() {
      return {
        gridTemplateAreas: createGridTemplate(this.grid).template,
      };
    },
    bodyClass() {
      return {
        padded: this.padded,
      };
    },
  },
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  @apply flex flex-col flex-grow transition duration-300 select-none p-px antialiased;
  background-color: #6277a244;

  &:hover {
    @apply duration-100;
  }
}

.header {
  @apply py-2 px-4 relative flex justify-center items-center;
  @apply uppercase leading-none font-bold;
  color: #fff8;
  font-size: 0.7rem;
}

.title {
  text-shadow: -1px -1px #0001;
}

.header-button {
  @apply absolute top-0 right-0 bottom-0 outline-none z-10 px-4;

  &.focus-visible {
    @apply shadow-outline;
  }

  &._toggled {
    @apply bg-gray-900 text-white;
  }

  &:hover {
    @apply text-white;
  }

  &:active {
    @apply bg-gray-800 transition-none text-white;
  }

  & .fa-icon {
    margin: -0.2em 0;
  }
}

.body-wrapper {
  @apply flex flex-grow relative items-stretch;
}

.body {
  @apply grid gap-px text-gray-400 flex-grow;
}

.body > * {
  @apply bg-gray-900;
}

.padded > * {
  @apply p-2;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  @apply transition duration-300;
}
.revealSettings-enter,
.revealSettings-leave-to {
  @apply opacity-0;
}
</style>
