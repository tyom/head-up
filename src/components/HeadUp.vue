<script setup>
import {
  ref,
  watchEffect,
  defineComponent,
  onMounted,
  onBeforeUnmount,
} from 'vue';
import { store } from '../store';
import LocalSidebar from './_Sidebar.vue';

defineComponent({
  components: {
    LocalSidebar,
  },
});

const storedHeadUpState = JSON.parse(
  window.localStorage.getItem('head-up') || '{}'
);

const sidebarVisible = ref(storedHeadUpState.sidebarVisible ?? true);
const smoothScroll = ref(false);
const activeBoardId = ref(window.location.hash.slice(1));
const activeBoardTitle = ref(null);

let activeBoardTitleTimeout;

function handleHashChange(evt) {
  const [, boardId] = evt.newURL.split('#');
  activeBoardId.value = boardId;
}

function handleSidebarToggle() {
  sidebarVisible.value = !sidebarVisible.value;
  window.localStorage.setItem(
    'head-up',
    JSON.stringify({
      sidebarVisible: sidebarVisible.value,
    })
  );
}

function displayBoardTitle(boardId, timeout = 1600) {
  const activeBoard = store.state.boards.find((x) => x.id === boardId);
  activeBoardTitle.value = activeBoard?.title;
  // self-destruct
  clearTimeout(activeBoardTitleTimeout);
  activeBoardTitleTimeout = setTimeout(() => {
    activeBoardTitle.value = null;
  }, timeout);
}

watchEffect(() => {
  displayBoardTitle(activeBoardId.value);
});

onMounted(() => {
  if (activeBoardId.value) {
    document.getElementById(activeBoardId.value)?.scrollIntoView();
  }
  smoothScroll.value = true;

  window.addEventListener('hashchange', handleHashChange);
});

onBeforeUnmount(() => {
  window.removeEventListener('hashchange', handleHashChange);
});
</script>

<template>
  <div class="head-up">
    <LocalSidebar
      :visible="sidebarVisible"
      :active-board="activeBoardId"
      :boards="store.getters.serialisedBoards"
      @toggle="handleSidebarToggle"
    />
    <main>
      <transition name="fade">
        <h2 v-if="activeBoardTitle" class="active-board-title">
          {{ activeBoardTitle }}
        </h2>
      </transition>
      <div class="boards" :class="{ '--smooth-scroll': smoothScroll }">
        <slot />
      </div>
    </main>
  </div>
</template>

<style scoped lang="postcss">
.head-up {
  @apply flex h-screen
    bg-gray-900 text-gray-100;

  & > * {
    @apply items-stretch;
  }
}

.active-board-title {
  @apply absolute top-1/2 left-1/2
    transform -translate-x-1/2 -translate-y-1/2
    py-5 px-10 rounded-full h-auto
    text-lg bg-black bg-opacity-60;
}

main {
  @apply relative flex-1;
}

.boards {
  @apply h-full overflow-hidden;

  &.--smooth-scroll {
    scroll-behavior: smooth;
  }

  & > ::v-deep(*) {
    @apply h-screen;
  }
}

.fade-enter-active,
.fade-leave-active {
  transition: 0.3s;
}
.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>
