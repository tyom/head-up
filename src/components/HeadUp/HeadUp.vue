<template>
  <div
    v-shortkey="shortkeys"
    :class="rootClass"
    class="HeadUp"
    @shortkey="handleKeys"
  >
    <Sidebar
      v-if="!hideSidebar"
      :visible="state.showSidebar"
      @toggle:settings="toggleSettingsScreen"
      @toggle:help="toggleHelpScreen"
    />
    <div class="main">
      <transition name="boardTitle">
        <div v-if="activeBoardTitle" class="board-title">
          {{ activeBoardTitle }}
        </div>
      </transition>
      <HeadUpBoards :boards="editableBoards" ref="boardContainer">
        <slot />
      </HeadUpBoards>
    </div>
    <ModalDialogue :content="modal" @close="modal = null" />
  </div>
</template>

<script>
import { get } from 'lodash';
import ally from 'ally.js';
import store from '../../store';
import { serializeSlot } from '../../transformers';
import ModalDialogue from '../ModalDialogue';
import Sidebar from '../Sidebar';
import HeadUpBoards from './HeadUpBoards';

export default {
  name: 'HeadUp',
  components: {
    HeadUpBoards,
    Sidebar,
    ModalDialogue,
  },
  props: {
    hideSidebar: {
      type: Boolean,
      default: false,
    },
    boards: {
      type: Array,
      default: () => [],
    },
  },
  data() {
    return {
      state: store.state,
      activeBoard: null,
      activeBoardTitle: null,
      activeBoardTitleTimeout: null,
      focusTrap: null,
      persistedState: null,
      modal: null,
      tabFocused: true,
      shortkeys: {
        up: ['k'],
        down: ['j'],
        sidebar: ['s'],
        edit: ['e'],
        add: ['a'],
        settings: ['shift', '<'],
        help: ['shift', '?'],
      },
    };
  },
  computed: {
    editableBoards() {
      return store.state.serializedBoards.filter(x => x.editable);
    },
    activeBoardIndex() {
      return store.state.serializedBoards.findIndex(
        x => x.id === this.state.activeBoardId,
      );
    },
    rootClass() {
      return {
        _edit: this.state.editMode,
        _paused: !this.tabFocused,
      };
    },
  },
  created() {
    store.dispatch('INIT_BOARDS', {
      fromProps: [...this.editableBoards, ...this.boards],
      fromSlots: serializeSlot(this.$slots.default),
    });
  },
  mounted() {
    // set up watchers on mount to get access to elements to scroll to
    this.setUpWatchers();

    const initialBoardId =
      this.state.activeBoardId || get(this.state, 'serializedBoards.[0].id');
    store.dispatch('ACTIVATE_BOARD', initialBoardId);

    window.addEventListener('focus', () => {
      this.tabFocused = true;
    });
    window.addEventListener('blur', () => {
      this.tabFocused = false;
    });
  },
  methods: {
    setUpWatchers() {
      this.$watch(
        'state.editMode',
        newVal => {
          newVal
            ? this.trapFocusInBoard(this.state.activeBoardId)
            : this.untrapFocus();
        },
        { immediate: true },
      );

      this.$watch(
        'state.activeBoardId',
        (newId, oldId) => {
          newId && this.activateBoard(newId, Boolean(oldId));
        },
        { immediate: true },
      );

      window.addEventListener('resize', () => {
        this.activateBoard(this.state.activeBoardId, false);
      });
    },
    activateBoard(boardId, smoothScroll = true) {
      this.scrollToBoard(boardId, smoothScroll);
      this.displayBoardTitle(boardId);

      if (this.state.editMode) {
        this.trapFocusInBoard(this.state.activeBoardId);
      }
    },
    scrollToBoard(boardId, smoothScroll) {
      const boardEl = this.$refs.boardContainer.$el.querySelector(
        `#board-${boardId}`,
      );

      if (!boardEl) {
        return;
      }

      const scrollOptions =
        smoothScroll && this.state.settings.smoothScrolling.value
          ? { behavior: 'smooth' }
          : {};
      boardEl.scrollIntoView(scrollOptions);
    },
    displayBoardTitle(boardId, timeout = 1600) {
      const board = this.state.serializedBoards.find(x => x.id === boardId);
      if (!board) {
        return;
      }
      this.activeBoardTitle = board.title;
      // self-destruct
      clearTimeout(this.activeBoardTitleTimeout);
      this.activeBoardTitleTimeout = setTimeout(() => {
        this.activeBoardTitle = null;
      }, timeout);
    },
    trapFocusInBoard(boardId) {
      this.untrapFocus();

      const boardEl = this.$refs.boardContainer.$el.querySelector(
        `#board-${boardId}`,
      );
      this.focusTrap = ally.maintain.tabFocus({
        context: boardEl,
      });
    },
    untrapFocus() {
      if (this.focusTrap) {
        this.focusTrap.disengage();
      }
    },
    handleKeys(evt) {
      switch (evt.srcKey) {
        case 'up':
          this.activatePreviousBoard();
          break;
        case 'down':
          this.activateNextBoard();
          break;
        case 'sidebar':
          store.dispatch('TOGGLE_SIDEBAR');
          break;
        case 'edit':
          store.dispatch('TOGGLE_EDIT_MODE');
          break;
        case 'add':
          store.dispatch('ADD_BOARD');
          break;
        case 'settings':
          this.toggleSettingsScreen();
          break;
        case 'help':
          this.toggleHelpScreen();
          break;
        case 'escape':
          this.closeModal();
          break;
        default:
      }
    },
    activateNextBoard() {
      const { serializedBoards } = this.state;
      const nextId =
        this.activeBoardIndex === serializedBoards.length - 1
          ? get(serializedBoards, '[0].id')
          : get(serializedBoards, `[${this.activeBoardIndex + 1}].id`);

      store.dispatch('ACTIVATE_BOARD', nextId);
    },
    activatePreviousBoard() {
      const { serializedBoards } = this.state;
      const prevId =
        this.activeBoardIndex === 0
          ? get(serializedBoards, `[${serializedBoards.length - 1}].id`)
          : get(serializedBoards, `[${this.activeBoardIndex - 1}].id`);

      store.dispatch('ACTIVATE_BOARD', prevId);
    },
    toggleHelpScreen() {
      this.toggleModal({ name: 'help', heading: 'Keyboard help' });
    },
    toggleSettingsScreen() {
      this.toggleModal({ name: 'settings', heading: 'Settings' });
    },
    toggleModal(modalData) {
      const existingModalName = this.modal && this.modal.name;
      this.modal = existingModalName === modalData.name ? null : modalData;
    },
  },
};
</script>

<style src="../../css/util.css" />

<style scoped>
>>> * {
  &,
  &::before,
  &::after {
    box-sizing: border-box;
    margin: 0;
  }
}

.HeadUp {
  font-size: responsive 8px 23px;
  transition: 0.4s background-color;
  background-color: #222;
  color: #bbb;
  height: 100vh;
  width: 100vw;
  display: flex;
  align-items: center;

  &._edit {
    background-color: #0d1a2b;
  }
}

.board-title {
  color: #fff;
  padding: 1em 2em;
  z-index: 1;
  background-color: #111a;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  border-radius: 2em;
}

.main {
  height: 100vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  position: relative;
}

.boardTitle-enter-active,
.boardTitle-leave-active {
  transition: 0.3s;
}
.boardTitle-enter,
.boardTitle-leave-to {
  opacity: 0;
}
</style>
