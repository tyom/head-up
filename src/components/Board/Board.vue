<template>
  <div :id="boardId" :class="boardClass" class="Board">
    <transition name="slideDown">
      <BoardToolbar
        v-if="!isThumb && editMode"
        :board="$props"
        @save="handleSave"
        @done="handleDone"
      />
    </transition>
    <div :class="layoutClass" class="cells">
      <template v-if="isThumb">
        <div v-for="(cell, idx) in cells" :key="idx" class="cell-placeholder" />
      </template>
      <template v-else>
        <slot v-if="slotCells" />
        <Cell
          v-else
          v-for="(cell, idx) in cells"
          v-bind="cell"
          :editable="editable"
          :key="idx"
          class="cell"
        />
      </template>
    </div>
  </div>
</template>

<script>
import get from 'lodash/get';
import store from '../../store';
import Cell from '../Cell';
import BoardToolbar from './BoardToolbar';

export default {
  name: 'Board',
  components: {
    Cell,
    BoardToolbar,
  },
  props: {
    id: {
      type: String,
      required: true,
    },
    isThumb: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: '',
    },
    cells: {
      type: Array,
      default: () => [],
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boardId() {
      return this.isThumb ? null : `board-${this.id}`;
    },
    boardClass() {
      return {
        _thumb: this.isThumb,
      };
    },
    editMode() {
      return store.state.editMode;
    },
    slotCells() {
      if (!this.$slots.default) {
        return;
      }
      const cells = this.$slots.default.filter(
        (x) => get(x, 'componentOptions.tag') === 'Cell',
      );
      return cells.length ? cells : undefined;
    },
    layoutClass() {
      const cells = this.slotCells || this.cells;
      if (!cells.length) {
        return;
      }
      return [`u-grid-${cells.length}-x`];
    },
  },
  methods: {
    handleSave(newState) {
      store.dispatch('UPDATE_BOARD', {
        boardId: this.id,
        payload: newState,
      });
    },
    handleDone(newState) {
      if (this.editable) {
        this.handleSave(newState);
      }
      store.dispatch('TOGGLE_EDIT_MODE', false);
    },
  },
};
</script>

<style scoped>
.Board {
  @apply flex flex-grow flex-col h-full;
}

.cell-placeholder {
  background-color: rgba(#fff, 0.1);
}

.cells {
  height: 100%;
  grid-gap: 0.5em;
  padding: 0.5em;
  overflow: auto;
}

._thumb {
  font-size: 0.8em;
  margin: auto;

  & .cells {
    padding: 2px;
    grid-gap: 2px;
  }

  & .cell {
    border: 0;
    background-color: #fff2;
  }
}

.slideDown-enter-active,
.slideDown-leave-active {
  transition: 0.1s;
  overflow: hidden;
}

.slideDown-enter,
.slideDown-leave-to {
  height: 0;
  opacity: 0;
}
</style>
