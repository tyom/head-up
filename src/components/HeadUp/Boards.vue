<template>
  <div class="HeadUpBoards flex-grow overflow-hidden">
    <Board
      v-for="(board, idx) in boards"
      v-bind="board"
      :key="board.title + board.id + idx"
    />
    <slot />
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
  },
};
</script>
