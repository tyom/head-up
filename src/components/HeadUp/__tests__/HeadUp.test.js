import { merge } from 'lodash';
import Vuex from 'vuex';
import ally from 'ally.js';
import ShortKey from 'vue-shortkey';
import Chance from 'chance';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import store, { initialState } from '../../../store';
import Board from '../../Board';
import Cell from '../../Cell';
import Sidebar from '../../Sidebar';
import ModalDialogue from '../../ModalDialogue';
import VSwitchToggle from '../../VSwitchToggle';
import HeadUp from '../HeadUp';
import HeadUpBoards from '../HeadUpBoards';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(ShortKey);
localVue.component('VSwitchToggle', VSwitchToggle);

function mountHeadUp(options = {}) {
  return shallowMount(HeadUp, {
    localVue,
    ...options,
    stubs: {
      Cell,
      Board,
      HeadUpBoards,
    },
  });
}

function mountWithSlot(options = {}) {
  return mountHeadUp(
    merge(
      {
        localVue,
        slots: {
          default: `
            <Board id="b1" title="Slot board">
              <Cell title="Cell #1"/>
            </Board>
          `,
        },
      },
      options,
    ),
  );
}

function mountWithProps(options = {}) {
  return mountHeadUp(
    merge(
      {
        localVue,
        propsData: {
          boards: [
            {
              id: '1',
              title: 'Prop Board #1',
              editable: true,
              cells: [{ title: 'Cell #1' }, { title: 'Cell #2' }],
            },
            {
              id: '2',
              title: 'Prop Board #2',
              editable: true,
              cells: [{ title: 'Cell #3' }],
            },
            {
              id: '3',
              title: 'Prop Board #3',
              editable: true,
              cells: [{ title: 'Cell #4' }],
            },
          ],
        },
      },
      options,
    ),
  );
}

beforeEach(() => {
  jest.clearAllMocks();
  store.replaceState({ ...initialState });
  Element.prototype.scrollIntoView = jest.fn();
  Chance.prototype.sentence = jest.fn(() => 'New board.');
});

describe('rendering', () => {
  test('render default', () => {
    const wrapper = mountHeadUp();
    expect(wrapper).toMatchSnapshot();
  });

  test('render without sidebar', () => {
    const wrapper = mountHeadUp({
      propsData: {
        hideSidebar: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  test('render in edit mode', () => {
    const wrapper = mountHeadUp();

    store.state.editMode = true;

    expect(wrapper).toMatchSnapshot();
  });

  test('render with slot', () => {
    const wrapper = mountWithSlot();

    expect(wrapper).toMatchSnapshot();
  });

  test('render with props', () => {
    const wrapper = mountWithProps();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('scrolling', () => {
  beforeEach(() => {
    mountWithProps();
  });

  test('initial scroll on mount', () => {
    expect(Element.prototype.scrollIntoView).toHaveBeenCalledTimes(1);
    expect(Element.prototype.scrollIntoView).toHaveBeenCalledWith({});
  });

  test('scroll on activation', () => {
    store.dispatch('ACTIVATE_BOARD', '2');

    expect(Element.prototype.scrollIntoView).toHaveBeenCalledTimes(2);
    expect(Element.prototype.scrollIntoView.mock.calls[1][0]).toEqual({
      behavior: 'smooth',
    });
  });

  test('invalid board id', () => {
    Element.prototype.scrollIntoView.mockClear();

    store.dispatch('ACTIVATE_BOARD', 'fake');

    expect(Element.prototype.scrollIntoView).not.toHaveBeenCalled();
  });
});

describe('focus trap', () => {
  ally.maintain.tabFocus = jest.fn(() => ({
    disengage: jest.fn(),
  }));

  let wrapper;

  beforeEach(() => {
    wrapper = mountWithProps();
  });

  test('trap focus', () => {
    expect(wrapper.vm.focusTrap).toBeNull();

    store.dispatch('TOGGLE_EDIT_MODE');

    expect(ally.maintain.tabFocus).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.focusTrap).toBeTruthy();
  });

  test('untrap focus', () => {
    store.dispatch('TOGGLE_EDIT_MODE');
    store.dispatch('TOGGLE_EDIT_MODE');

    expect(wrapper.vm.focusTrap.disengage).toHaveBeenCalledTimes(1);
  });
});

describe('manipulations', () => {
  test('add board', () => {
    const wrapper = mountWithProps();

    expect(wrapper.vm.state.serializedBoards.length).toEqual(3);

    store.dispatch('ADD_BOARD');

    expect(store.state.serializedBoards.length).toEqual(4);
    expect(store.state.serializedBoards[0].title).toEqual('new board');
    expect(store.state.serializedBoards[0].id).toEqual('new-board');

    expect(Element.prototype.scrollIntoView).toHaveBeenCalledWith({});
  });

  test('remove active board', () => {
    const wrapper = mountWithProps();

    expect(store.state.serializedBoards.length).toEqual(3);
    expect(wrapper.vm.state.activeBoardId).toEqual('1');

    localVue.nextTick(() => {
      store.dispatch('REMOVE_BOARD', '1');

      expect(store.state.serializedBoards.length).toEqual(2);
      expect(wrapper.vm.state.activeBoardId).toEqual('2');
      expect(wrapper.vm.activeBoardIndex).toEqual(0);

      expect(Element.prototype.scrollIntoView).toHaveBeenCalledWith({
        behavior: 'smooth',
      });
    });
  });

  test('remove board preceding the active one', () => {
    const wrapper = mountWithProps();

    localVue.nextTick(() => {
      wrapper.vm.state.activeBoardId = '2';

      expect(store.state.serializedBoards.length).toEqual(3);
      expect(wrapper.vm.state.activeBoardId).toEqual('2');

      store.dispatch('REMOVE_BOARD', '1');

      expect(store.state.serializedBoards.length).toEqual(2);
      expect(wrapper.vm.state.activeBoardId).toEqual('2');
      expect(wrapper.vm.activeBoardIndex).toEqual(0);

      expect(Element.prototype.scrollIntoView).toHaveBeenCalledWith({
        behavior: 'smooth',
      });
    });
  });

  test('remove last active board in the list', () => {
    const wrapper = mountWithProps();

    localVue.nextTick(() => {
      wrapper.vm.state.activeBoardId = '3';

      expect(store.state.serializedBoards.length).toEqual(3);
      expect(wrapper.vm.state.activeBoardId).toEqual('3');

      store.dispatch('REMOVE_BOARD', '3');

      expect(store.state.serializedBoards.length).toEqual(2);
      expect(wrapper.vm.state.activeBoardId).toEqual('2');
      expect(wrapper.vm.activeBoardIndex).toEqual(1);

      expect(Element.prototype.scrollIntoView).toHaveBeenCalledWith({
        behavior: 'smooth',
      });
    });
  });
});

describe('board editing', () => {
  const newBoardData = {
    id: '1',
    title: 'Updated board',
  };

  beforeEach(() => {
    mountWithProps();
  });

  test('apply edits', () => {
    store.dispatch('UPDATE_BOARD', {
      boardId: '1',
      payload: newBoardData,
    });

    const editedBoard = store.state.serializedBoards.find(x => x.id === '1');
    expect(editedBoard).toHaveProperty('title', newBoardData.title);
  });
});

describe('modal dialogues', () => {
  let wrapper;
  let sidebar;

  beforeEach(() => {
    wrapper = mountWithSlot();
    sidebar = wrapper.find(Sidebar);
  });

  test('open modal', () => {
    sidebar.vm.$emit('toggle:help');
    expect(wrapper.vm.modal.name).toEqual('help');
    expect(wrapper).toMatchSnapshot();

    sidebar.vm.$emit('toggle:help');
    expect(wrapper.vm.modal).toBeNull();
  });

  test('settings', () => {
    sidebar.vm.$emit('toggle:settings');
    expect(wrapper.vm.modal.name).toEqual('settings');
    expect(wrapper).toMatchSnapshot();

    wrapper.find(ModalDialogue).vm.$emit('close');
    expect(wrapper.vm.modal).toBeNull();
  });
});

describe('respond to keystrokes', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithProps({
      attachToDocument: true,
    });
  });

  test('next board', () => {
    expect(wrapper.vm.state.activeBoardId).toEqual('1');
    wrapper.trigger('keydown', {
      key: 'j',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('2');
    wrapper.trigger('keydown', {
      key: 'j',
    });
    wrapper.trigger('keydown', {
      key: 'j',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('1');
  });

  test('previous board', () => {
    wrapper.trigger('keydown', {
      key: 'k',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('3');

    wrapper.trigger('keydown', {
      key: 'k',
    });
    expect(wrapper.vm.state.activeBoardId).toEqual('2');
  });

  test('toggle sidebar', () => {
    expect(wrapper.vm.state.showSidebar).toEqual(true);
    wrapper.trigger('keydown', {
      key: 's',
    });
    expect(wrapper.vm.state.showSidebar).toEqual(false);
  });

  test('toggle edit', () => {
    expect(wrapper.vm.state.editMode).toEqual(false);
    wrapper.trigger('keydown', {
      key: 'e',
    });
    expect(wrapper.vm.state.editMode).toEqual(true);
  });

  test('add board', () => {
    expect(wrapper.vm.state.serializedBoards.length).toEqual(3);
    wrapper.trigger('keydown', {
      key: 'a',
    });
    expect(wrapper.vm.state.serializedBoards.length).toEqual(4);
  });

  //  TODO: Keystroke combinations
});

test('reactivate board on window resize', () => {
  const wrapper = mountWithSlot({
    attachToDocument: true,
  });
  jest.spyOn(wrapper.vm, 'activateBoard');

  const resizeEvent = new Event('resize');
  global.dispatchEvent(resizeEvent);

  expect(wrapper.vm.activateBoard).toHaveBeenCalledWith('b1', false);
});
