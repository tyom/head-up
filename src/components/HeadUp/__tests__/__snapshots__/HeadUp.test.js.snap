// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal dialogues open modal 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <div class="board-title">
        Slot board
      </div>
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden">
      <div id="board-b1" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`modal dialogues settings 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <div class="board-title">
        Slot board
      </div>
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden">
      <div id="board-b1" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`rendering render default 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <!---->
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden"> </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`rendering render in edit mode 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <!---->
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden"> </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`rendering render with props 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <!---->
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden">
      <div id="board-1" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-2-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #2</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
      <div id="board-2" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-1-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #3</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
      <div id="board-3" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-1-x">
          <div class="Cell cell">
            <header class="header">
              <h2 class="title">Cell #4</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`rendering render with slot 1`] = `
<div class="HeadUp">
  <sidebar-stub visible="true"></sidebar-stub>
  <div class="main">
    <transition-stub name="boardTitle">
      <!---->
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden">
      <div id="board-b1" class="Board">
        <transition-stub name="slideDown">
          <!---->
        </transition-stub>
        <div class="cells u-grid-1-x">
          <div class="Cell">
            <header class="header">
              <h2 class="title">Cell #1</h2>
              <!---->
            </header>
            <div class="body-wrapper">
              <div class="body"></div>
              <transition-stub name="revealSettings">
                <!---->
              </transition-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;

exports[`rendering render without sidebar 1`] = `
<div class="HeadUp">
  <!---->
  <div class="main">
    <transition-stub name="boardTitle">
      <!---->
    </transition-stub>
    <div class="HeadUpBoards flex-grow overflow-hidden"> </div>
  </div>
  <modaldialogue-stub></modaldialogue-stub>
</div>
`;
